### ISSUE TITLE
Reduce or Eliminate getMasterCs and setMasterCs

### ISSUE THREAD

**Issue by john-science** [terrapower/armi#930]
The `getMasterCs()` and `setMasterCs()` methods are extremely problematic. First off, obviously, they prevent ARMI from being able to do multiple things at once. We can not do the following currently, because they mean over-writing the "master cs":

* read two ARMI DBs at one time
* have two ARMI Reactors in memory at once
* All unit tests must share the same ARMI `App()`
* any `Block`s in memory must reference the same settings file
* All interfaces have access to their own settings object (`self.cs`), but can also reference the "master cs"

Here are all locations of usage:

### getMasterCs

* [x] `armi/bookkeeping/report/html.py`:    `cs = settings.getMasterCs()`
* [ ] `armi/interfaces.py`:        `self.cs = settings.getMasterCs() if cs is None else cs`
* [ ] `armi/interfaces.py`:        `self.cs = cs or settings.getMasterCs()`
* [ ] `armi/interfaces.py`:        `self.cs = settings.getMasterCs()`
* [ ] `armi/operators/operatorMPI.py`:        `cs = settings.getMasterCs()`
* [x] `armi/reactor/blocks.py`:        `if len(xsType) == 2 and len(settings.getMasterCs()["buGroups"]) == 1:`
* [x] `armi/reactor/blocks.py`:        `burnupPeakingFactor = settings.getMasterCs()["burnupPeakingFactor"]`
* [x] `armi/reactor/blocks.py`:        `if settings.getMasterCs()["gridPlateDpaXsSet"] and self.hasFlags(`
* [x] `armi/reactor/blocks.py`:            `dpaXsSetName = settings.getMasterCs()["gridPlateDpaXsSet"]`
* [x] `armi/reactor/blocks.py`:            `dpaXsSetName = settings.getMasterCs()["dpaXsSet"]`
* [x] `armi/reactor/reactors.py`:        `self.processLoading(settings.getMasterCs())`
* [x] `armi/reactor/reactors.py`:        `self.buildZones(settings.getMasterCs())`
* [x] `armi/reactor/reactors.py`:            `cs or settings.getMasterCs(), name=assemType`
* [ ] `armi/operators/tests/test_operators.py`:            `self.cs = settings.getMasterCs()`

If we can remove all the instances of `getMasterCs()`, we will be able to just delete all `setMasterCs()` for free.

**Comment by jakehader** [comment]
This is a tricky one but I agree! I think that one solution could be to pass all settings into where they are needed, but this might be messy. Maybe we can resolve settings to parameter data on the reactor core model? As in, this reactor model was generated in the context of these settings?

**Comment by john-science** [comment]
@jakehader I think there will be a mix of solutions, but that is definitely one of the big ones.

For instance, the reporting system currently takes the entire `CaseSetings` object through a `getMasterCs()`, but [all it actually uses is the case title](https://github.com/terrapower/armi/blob/5db113423d0981f8887809b3297b97e3117265e6/armi/bookkeeping/report/html.py#L210). That's just one string we could pass in. And, frankly, it only uses it to make a plot look pretty, so it's more of a nice-to-have than important to the math.

**Comment by john-science** [comment]
@jakehader Oh, I just realized you were talking about the `Reactor`, my bad.

Yeah, right now the `Operator` has a `CaseSetting` object, but not the `Reactor` or `Core`. We _could_ give the `Reactor` access to a `self.cs`, but I think there is an easier way.

Right now, the `Reactor` never uses the `cs` object, but `Core` does in three places:

1. `buildZones()`.  Well, Josh Chen and I have a `Zones` branch open that just happens to fix that.  Done.
2. We call `processLoading(cs)` inside `removeAssembliesInRing()`. Well, maybe that `processLoading()` should really be called in whatever code calls `removeAssembliesInRing(), right?  Easy  (Also, I note, `removeAssembliesInRing()` is not _used_ anywhere. We could delete it and no one would notice.)
3. `createAssemblyOfType()` is only called from `geometryConverters.py` and `fuelHandlers`.py`, which both have a `self.cs`.

**TLDR**; I think the solution for `reactors.py` is not to make a high-level change where we _give_ `Reactor` or `Core` their own `self.cs`, but to look at these 3 uses of `getMasterCs()` as exceptions and mistakes, and just fix them.

**Comment by keckler** [comment]
Out of the options that you suggest, I would say option (4) is the most ideal.
I think my second choice would be option (2).

But I would also suggest another possibility, maybe it's good, maybe it's not.

5. Change the logic to only complain when a second burnup group actually exists:
```python
if len(sxsType) == 1:
    return xsType + bu
elif len(xsType) == 2 and ord(bu) < ord("A"):
    raise ValueError(f"Use of multiple burnup groups is not allowed with multi-character xs groups!")
else:
    return xsType
```

To me, something like that seems like it could work without having to pass in any extra information. It has the drawback of potentially failing late in the simulation, though. We could warn against that somewhere in the docs, perhaps.

**Comment by john-science** [comment]
@keckler I like the idea. It seems much cleaner and easier to support to me. I will have to play with it. If it works, I think that's a great solution.

**Comment by john-science** [comment]
Okay, all that is left is the `Interface` and `OperatorMPI`.

Most of the `Interface` items have the option of passing in a `CaseSetting` object anyway. So it might be as easy as squashing all the times where we aren't passing in a CS object manually.

I do not yet understand the logical flow is for the use of `getMasterCs()` is in `OperatorMPI._resetWorker()`, especially when `OperatorMPI` inherits from `Operator`, and those has a `self.cs` in it. Because MPI means multiple "processes", so the different processes don't share the same "master CS" object.  Why not just get it from `self.cs`?  Either I am missing something (possible), or this was never a necessary call to `getMasterCs()` to begin with.
